### ISSUE TITLE
package.dhall

### ISSUE THREAD

**Issue by f-f** [arianvp/dhall-kubernetes#11]
The Prelude has a [`package.dhall`](https://github.com/dhall-lang/Prelude/blob/master/package.dhall) file (in the root and every subdirectory) that allows you to import the whole package once and access the various components via record selector.

Example:
```haskell
-- To get the List/map function
let Prelude = https://raw.githubusercontent.com/dhall-lang/Prelude/master/package.dhall
in Prelude.`List`.map
```

I would like to do something similar for our repo, but there would be different possibilities:
1. package file in the root would be similar to the Prelude one: `{ types = ./types/package.dhall, default = ./default/package.dhall }`.
  In this way to access e.g. a Deployment type and default, one would do
  ```haskell
     let kube = https://raw.githubusercontent.com/dhall-lang/dhall-kubernetes/master/package.dhall
  in let DeploymentType = kube.types.io.k8s.api.apps.v1beta2.Deployment.dhall
  in let DeploymentDefault = kube.default.io.k8s.api.apps.v1beta2.Deployment.dhall
  in DeploymentDefault : DeploymentType 
  -- actually this would not typecheck (DeploymentDefault is a function), but it's for demonstration purposes
  ```
2. package file in the root would have keys as the Kube entities, and values as a record with type and default: 
  ```haskell
  { io.k8s.api.apps.v1beta2.Deployment =
    { type = ./types/io.k8s.api.apps.v1beta2.Deployment.dhall
    , default = ./default/io.k8s.api.apps.v1beta2.Deployment.dhall
    }
  , ...
  }
  ```
  In this way to access e.g. a Deployment type and default, one would do
  ```haskell
     let kube = https://raw.githubusercontent.com/dhall-lang/dhall-kubernetes/master/package.dhall
  in let Deployment = kube.io.k8s.api.apps.v1beta2.Deployment
  in Deployment.default : Deployment.type 
  -- actually this would not typecheck (DeploymentDefault is a function), but it's for demonstration purposes
  ```

/cc @arianvp opinions? (I'm leaning towards 2, as it's less lines of boilerplate for the user)

PS: I'm not sure we could use dots in record keys though, so we might need to go back to the nested directories layout, but it's a detail here

**Comment by arianvp** [comment]
I much prefer option 2, it seems a lot more ergonomic.  However, I don't think  dots are supported in record keys though? Maybe it is?

In nix the following:

```
{ a.b.c.d = 3 }
```

is syntactic sugar for

```
{ a = { b = { c  = { d = 3 }}}}} 
```

Is it the same in Dhall?

Then this should work as expected right?

**Comment by f-f** [comment]
@arianvp turns out we cannot mix types and values in the same record (as I mentioned also in https://github.com/dhall-lang/dhall-kubernetes/issues/18#issuecomment-410801036), so we cannot go with option 2: only option 1 with two separate files (types and defaults) is viable.

**Comment by arianvp** [comment]
ah that's a shame.

**Comment by Profpatsch** [comment]
https://github.com/dhall-lang/dhall-lang/issues/219 would fix that problem and not require records.

**Comment by f-f** [comment]
@Profpatsch how would it fix it without records?

Also I opened this issue to fix the problem of having a huge list of imports at the beginning of files (e.g. see [here](https://github.com/dhall-lang/dhall-kubernetes/blob/master/examples/deployment.dhall)), but I think a more fitting solution would be to fix the problem with #18 instead: the idea is the same, but this `package.dhall` is meant to be project-wide (which is not that useful actually, who would use all the entities in the same file anyways?) while that one is focused on a specific use-case (so much more realistic on real-world usage).

**Comment by f-f** [comment]
Now that #34 got in, there is basically two records per Kubernetes object (e.g. see here the [types](./api/Deployment/RawTypes) and [defaults](./api/Deployment/RawDefaults) for a Deployment).
So IMO there's no need for a global `package.dhall` anymore - since one can just import them as needed - and we can close this issue
